dll/win32/shlwapi/utils.cpp

```diff
@@ -35,65 +35,174 @@
 
 WINE_DEFAULT_DEBUG_CHANNEL(shell);
 
-/*************************************************************************
- * IContextMenu_Invoke [SHLWAPI.207]
- *
- * Used by Win:SHELL32!CISFBand::_TrySimpleInvoke.
- */
-EXTERN_C
-BOOL WINAPI
-IContextMenu_Invoke(
-    _In_ IContextMenu *pContextMenu,
-    _In_ HWND hwnd,
-    _In_ LPCSTR lpVerb,
-    _In_ UINT uFlags)
+static inline WORD 
+GetVersionMajorMinor()
+{
+    DWORD version = GetVersion();
+    return MAKEWORD(HIBYTE(version), LOBYTE(version));
+}
+
+static HRESULT
+SHInvokeCommandOnContextMenuInternal(
+    _In_opt_ HWND hWnd,
+    _In_opt_ IUnknown* pUnk,
+    _In_ IContextMenu* pCM,
+    _In_ UINT fCMIC,
+    _In_ UINT fCMF,
+    _In_opt_ LPCSTR pszVerb,
+    _In_opt_ LPCWSTR pwszDir,
+    _In_ bool ForceQCM)
 {
-    CMINVOKECOMMANDINFO info;
-    BOOL ret = FALSE;
+    CMINVOKECOMMANDINFOEX info = { sizeof(info), fCMIC, hWnd, pszVerb };
     INT iDefItem = 0;
     HMENU hMenu = NULL;
     HCURSOR hOldCursor;
+    HRESULT hr = S_OK;
+    WCHAR wideverb[MAX_PATH];
 
-    TRACE("(%p, %p, %s, %u)\n", pContextMenu, hwnd, debugstr_a(lpVerb), uFlags);
-
-    if (!pContextMenu)
-        return FALSE;
+    if (!pCM)
+        return E_INVALIDARG;
 
     hOldCursor = SetCursor(LoadCursorW(NULL, (LPCWSTR)IDC_WAIT));
-
-    ZeroMemory(&info, sizeof(info));
-    info.cbSize = sizeof(info);
-    info.hwnd = hwnd;
     info.nShow = SW_NORMAL;
-    info.lpVerb = lpVerb;
+    if (pUnk)
+        IUnknown_SetSite(pCM, pUnk);
 
-    if (IS_INTRESOURCE(lpVerb))
+    if (IS_INTRESOURCE(pszVerb))
     {
         hMenu = CreatePopupMenu();
         if (hMenu)
         {
-            pContextMenu->QueryContextMenu(hMenu, 0, 1, MAXSHORT, uFlags | CMF_DEFAULTONLY);
+            hr = pCM->QueryContextMenu(hMenu, 0, 1, MAXSHORT, fCMF | CMF_DEFAULTONLY);
             iDefItem = GetMenuDefaultItem(hMenu, 0, 0);
             if (iDefItem != -1)
                 info.lpVerb = MAKEINTRESOURCEA(iDefItem - 1);
         }
+        info.lpVerbW = MAKEINTRESOURCEW(info.lpVerb);
     }
+    else
+    {
+        if (GetVersionMajorMinor() >= _WIN32_WINNT_WIN7)
+        {
+            info.fMask |= CMF_OPTIMIZEFORINVOKE;
+        }
+        if (pszVerb && SHAnsiToUnicode(pszVerb, wideverb, _countof(wideverb)))
+        {
+            info.fMask |= CMIC_MASK_UNICODE;
+            info.lpVerbW = wideverb;
+        }
+        if (ForceQCM)
+        {
+            hMenu = CreatePopupMenu();
+            hr = pCM->QueryContextMenu(hMenu, 0, 1, MAXSHORT, fCMF);
+        }
+    }
+
+    SetCursor(hOldCursor);
 
-    if (iDefItem != -1 || info.lpVerb)
+    if (SUCCEEDED(hr) && (iDefItem != -1 || info.lpVerb))
```

Comment:
There is completely no error / debug logging in this function left,
I think it would make sense to at least use `!FAILED_UNEXPECTEDLY(hr)` here.

Reply:
Unfortunately it's not easy to include shellutils.h here, gives me errors about ATL even if I include atlbase.h and atlcom.h first. I'll just add in a FAILED_UNEXPECTEDLY without logging for now...